src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -304,6 +309,163 @@ CreateDistributedInsertSelectPlan(Query *originalQuery,
 }
 
 
+/*
+ * CreateInsertSelectIntoLocalTablePlan creates the plan for INSERT .. SELECT queries
+ * where the selected table is distributed and the inserted table is not.
+ *
+ * To create the plan, this function first creates a distributed plan for the SELECT
+ * part. Then puts it as a subquery to the original (non-distributed) INSERT query as
+ * a subquery. Finally, it puts this INSERT query, which now has a distributed SELECT
+ * subquery, in the masterQuery.
+ *
+ * If the SELECT query is a router query, whose distributed plan does not have a
+ * masterQuery, this function also creates a dummy masterQuery for that.
+ */
+DistributedPlan *
+CreateInsertSelectIntoLocalTablePlan(uint64 planId, Query *originalQuery, ParamListInfo
+									 boundParams, bool hasUnresolvedParams,
+									 PlannerRestrictionContext *plannerRestrictionContext)
+{
+	RangeTblEntry *selectRte = ExtractSelectRangeTableEntry(originalQuery);
+
+	Query *selectQuery = BuildSelectForInsertSelect(originalQuery);
+	originalQuery->cteList = NIL;
+	DistributedPlan *distPlan = CreateDistributedPlan(planId, selectQuery,
+													  copyObject(selectQuery),
+													  boundParams, hasUnresolvedParams,
+													  plannerRestrictionContext);
+	if (distPlan->planningError)
```

Comment:
Maybe add comment (and maybe Assert) that we don't expect NULL here because we already checked hasUnresolvedParams prior to this function and CreateDistributedPlan only returns NULL when there are unresolved parameters.